node/src/components/consensus/highway_core/state.rs

```diff
@@ -705,6 +769,29 @@ impl<C: Context> State<C> {
     pub(crate) fn is_empty(&self) -> bool {
         self.units.is_empty()
     }
+
+    /// Returns the panorama of the confirmation for the leader unit `vhash`.
+    pub(crate) fn confirmation_panorama(
+        &self,
+        vidx: ValidatorIndex,
+        vhash: &C::Hash,
+    ) -> Panorama<C> {
+        let unit = self.unit(vhash);
+        let mut panorama;
+        // TODO(HWY-167): Confirmation panorama.
+        if let Some(prev_hash) = self.panorama().get(vidx).correct().cloned() {
+            let own_unit = self.unit(&prev_hash);
+            panorama = unit.panorama.merge(self, &own_unit.panorama);
+            panorama[vidx] = Observation::Correct(prev_hash);
+        } else {
+            panorama = unit.panorama.clone();
+        }
+        panorama[unit.creator] = Observation::Correct(*vhash);
+        for faulty_v in self.faulty_validators() {
+            panorama[faulty_v] = Observation::Faulty;
+        }
+        panorama
```

Comment:
I wonder whether we should simplify this, because it will be pretty difficult to merge the panoramas and still make sure we're LNC-compliant:

We could just check whether `vhash` (which should probably be `uhash` now) sees our own most recent unit: If it does, use `vhash`'s panorama and only update `vhash` itself; if it doesn't maybe don't create the confirmation at all. If the leader didn't see us, we pretend we didn't see the leader. That doesn't break the liveness proof. (But should be discussed with the team anyway, I guess).

Reply:
You might be right, I don't know yet, but this PR isn't really about implementing the logic for creating LNC-complaint units yet. Just moving code around while trying to preserve the old logic.